staging/src/k8s.io/client-go/tools/pager/pager.go

```diff
@@ -194,9 +194,15 @@ func (p *ListPager) eachListChunkBuffered(ctx context.Context, options metav1.Li
 		}()
 		err = p.eachListChunk(ctx, options, func(chunk runtime.Object) error {
 			select {
-			case chunkC <- chunk: // buffer the chunk, this can block
 			case <-ctx.Done():
```

Comment:
Hmm, thinking a bit more, this means that if the context expired _after_ the request completed but _before_ we called the callback (e.g. while decode was running), we would now return a timeout error instead of calling the callback. Is that what we want?

Reply:
Isn't that the behavior that this test implies (which is why it flakes without this change):

https://github.com/kubernetes/kubernetes/blob/661899f1a14680840e06c84bc3ff46f73d203012/staging/src/k8s.io/client-go/tools/pager/pager_test.go#L301-L307

Reply:
> Isn't that the behavior that this test implies (which is why it flakes without this change):

probably so... is that the behavior we want? will clients that currently process a change whose request completed within the timeout but whose callback was not called within the timeout think this fix is good or bad?

Reply:
it seems the context has preference in all the other methods here L87, L227, and is only way to do this method cancellable, right? the callback doesn't respect the context, in a edge case it may never return

Reply:
I think even with this change the callback can always hang forever because it is run in the main go routine.

Reply:
the callback is passed in along with the context, so if it's doing anything blocking it can cancel on the context itself, but you're right, we can't assume it's going to check/honor the context

Reply:
> I think even with this change the callback can always hang forever because it is run in the main go routine.

yeah, that is unavoidable, I mean that at least,with this change, the behavior is predictable, context wins and cancels, before it was a 50% chance when context and callback arrives at the same time

Reply:
also, the channel is buffered, it means it is possible some are going to go through

Reply:
sigh... I didn't realize chunkC was buffered... @enj, that means your first location for this select (inside the `for o := range chunkC` block) was correct. Sorry 🤦

Reply:
Restored back to the original SHA.